katsdpingest/ingest_session.py

```diff
@@ -1546,6 +1628,64 @@ def close(self) -> None:
             del self.sd_input_resource
             del self.sd_output_resource
 
+    async def _watch_sensor(self, telstate: katsdptelstate.aio.TelescopeState, key: str,
+                            notify: Callable[[Any, float], None]) -> None:
+        """Subscribe to updates for a sensor.
+
+        Each time the sensor is updated with a newer timestamp, the `notify`
+        callback is called. If an update is received for an earlier point in
+        time it is ignored. If there is already data in the sensor when called,
+        the last value is immediately notified.
+        """
+        ts = None
+
+        def callback(value: Any, new_ts: Optional[float]) -> bool:
+            nonlocal ts
+            if new_ts is None:
+                logger.warning('%s was set to an immutable value, should be mutable', key)
+                return False
+            if ts is None or new_ts > ts:
+                ts = new_ts
+                notify(value, ts)
+            else:
+                logger.warning('%s timestamp went backwards (%.3f < %.3f)', new_ts, ts)
+            return False
+
+        # The callback always returns false, so this call to wait_key will
+        # never return. Instead, it will be cancelled when the generator is
+        # shut down.
+        await telstate.wait_key(key, callback)
```

Comment:
I sort of expected the callback in async telstate to be async too... I guess it should be a quick routine in general anyway.

Reply:
I guess we could consider that in future if needed (e.g. to compare to some other data pulled from telstate) - but for the common case the asynchronous work of getting the value for this key from telstate has already been done.